Implement basic CanvasRenderingContext2D layers support without filters

tuankiet65 opened the pull request:
#### 69cc99cde3eb5473a077ac51c2c4ea064cbc2a69
<pre>
Implement basic CanvasRenderingContext2D layers support without filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=259784">https://bugs.webkit.org/show_bug.cgi?id=259784</a>
rdar://113343677

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-reset-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-save-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.save-beginLayer-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.save-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-save-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.save-beginLayer-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::contextIs2DBaseWithUnclosedLayers const):
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::toDataURL):
(WebCore::HTMLCanvasElement::toBlob):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::convertToBlob):
* Source/WebCore/html/canvas/BeginLayerOptions.h: Added.
* Source/WebCore/html/canvas/BeginLayerOptions.idl: Added.
* Source/WebCore/html/canvas/CanvasLayers.idl: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::unwindStateStack):
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::State::State):
(WebCore::CanvasRenderingContext2DBase::realizeSavesLoop):
(WebCore::CanvasRenderingContext2DBase::restore):
(WebCore::CanvasRenderingContext2DBase::clearRect):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::createPattern):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
(WebCore::CanvasRenderingContext2DBase::beginLayer):
(WebCore::CanvasRenderingContext2DBase::endLayer):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::hasOpenLayers const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69cc99cde3eb5473a077ac51c2c4ea064cbc2a69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16252 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16571 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18019 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15248 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16684 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17636 "13 flakes 162 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16892 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18785 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14135 "2 flakes 1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14713 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21541 "5 flakes 125 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14000 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15124 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18138 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15490 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15471 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13115 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/16455 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14695 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4095 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19061 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/17622 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15295 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3887 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

tuankiet65 commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -264,6 +264,10 @@ bool CanvasRenderingContext2DBase::isAccelerated() const
 {
+    // Unwind layers
+    while (m_layerCount--)
+        endLayer();
```
This is wrong, it will not work with intertwined beginLayer()/save().

weinig commented on Tools/DumpRenderTree/TestOptions.cpp:
```diff
@@ -67,6 +67,7 @@ const TestFeatures& TestOptions::defaults()
             // These are non-experimental WebPreference values that must always be set as they
             // differ from the default set in the WebPreferences*.yaml configuration.
             { "AllowsInlineMediaPlayback", true },
+            { "Canvas2DLayersEnabled", true },
```
Why does this need to be set? New prefs really should not be using this.

tuankiet65 replied:
I didn't follow the new developments so what's the proper way to do this? I don't want to enable this pref by default, but still have it on during layout tests

weinig commented on Tools/WebKitTestRunner/TestOptions.cpp:
```diff
@@ -80,6 +80,7 @@ const TestFeatures& TestOptions::defaults()
             { "AsyncOverflowScrollingEnabled", false },
             { "BroadcastChannelOriginPartitioningEnabled", false },
             { "BuiltInNotificationsEnabled", false },
+            { "Canvas2DLayersEnabled", true },
```
Same here.

tuankiet65 commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2857,4 +2914,63 @@ DestinationColorSpace CanvasRenderingContext2DBase::colorSpace() const
+
+    m_layerCount--;
+
+    didDrawEntireCanvas();
```
maybe should just call didDraw() with the clipping region prior to beginLayer()? need tests

tuankiet65 commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -206,6 +207,13 @@ template<typename T> static Ref<JSON::ArrayOf<JSON::Value>> buildArrayForVector(
+std::optional<InspectorCanvasCallTracer::ProcessedArgument> InspectorCanvas::processArgument(BeginLayerOptions argument)
+{
+    // FIXME: implement this.
+    (void)(argument);
```
ask Patrick about this.

tuankiet65 commented on Source/WebCore/html/canvas/CanvasLayers.idl:
```diff
@@ -0,0 +1,35 @@
+]
+interface mixin CanvasLayers {
+    // layers
+    undefined beginLayer(optional BeginLayerOptions options = {});
```
get rid of this for this patch

tuankiet65 commented on Source/WebCore/html/canvas/BeginLayerOptions.h:
```diff
@@ -0,0 +1,38 @@
+#include <any>
+
+#include <wtf/Vector.h>
+#include <wtf/text/WTFString.h>
```
unused imports.

heycam commented on LayoutTests/TestExpectations:
```diff
@@ -6615,3 +6611,34 @@ webkit.org/b/225349 imported/w3c/web-platform-tests/html/canvas/offscreen/reset/
 # We only support import-attributes and do not support import-assertions.
 imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions
+
+# CanvasRenderingContext2D Layers tests that uses dictionary filters, which we do not support.
```
"tests that use"

heycam commented on LayoutTests/TestExpectations:
```diff
@@ -6615,3 +6611,34 @@ webkit.org/b/225349 imported/w3c/web-platform-tests/html/canvas/offscreen/reset/
+webkit.org/b/225349 imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.reset.html [ ImageOnlyFailure ]
+
+# OffscreenCanvas does not support the new layers API right now.
+webkit.org/b/259784 imported/w3c/web-platform-tests/html/canvas/offscreen/layers [ Skip ]
```
Why is this? I would think that it should just work on offscreen canvases. Please file a followup bug to get this working if it's not.

heycam commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -381,6 +382,20 @@ void CanvasBase::resetGraphicsContextState() const
+        return false;
+    const auto& context2DBase = downcast<CanvasRenderingContext2DBase>(context);
+
+    return context2DBase.layerCount() > 0;
```
I see, for example, that `HTMLCanvasElement::clearImageBuffer` does a check and downcasts to `CanvasRenderingContext2D`, so I think it would be slightly better to put this logic in `HTMLCanvasElement`. Let's say a new function `HTMLCanvasElement::hasUnclosedLayers()`, which checks if the context is a `CanvasRenderingContext2D` and if so looks at its layer count.

tuankiet65 replied:
I think that this method should be in `CanvasBase`, since it is used by both `HTMLCanvasElement` and `OffscreenCanvas`.

heycam replied:
OK that's a good reason.

heycam commented on Source/WebCore/html/canvas/BeginLayerOptions.idl:
```diff
@@ -0,0 +1,32 @@
+
+[
+    EnabledBySetting=Canvas2DLayersEnabled
+]
```
It's fine to leave this but I think it's unnecessary, since IDL dictionaries don't get exposed in any way to the page (it's just the IDL attributes and operations etc. that do).

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -235,14 +235,23 @@ CanvasRenderingContext2DBase::CanvasRenderingContext2DBase(CanvasBase& canvas, C
+        return;
+
+    while (m_stateStack.size() > 1) {
+        State state = m_stateStack.takeLast();
```
Nit: use `auto` instead of `State`, since we generally avoid writing the types out unless it's unobvious or confusing.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -235,14 +235,23 @@ CanvasRenderingContext2DBase::CanvasRenderingContext2DBase(CanvasBase& canvas, C
+
+        context->restore();
     }
 }
```
Maybe assert at the end of the function that `m_layerCount` is 0? Then you could drop the assertion before `m_layerCount--` since it would cover that.

tuankiet65 replied:
The point here is we don't want m_layerCount() to go below 0 and underflow. Having one assertion at the end should be enough but I rather we don't decrement beyond zero in the first place.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -450,29 +461,41 @@ void CanvasRenderingContext2DBase::realizeSavesLoop()
-    c->restore();
+    if (c)
+        c->restore();
+
```
Could write this as:

```
if (auto* c = drawingContext())
    c->restore();
```

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
```diff
@@ -141,7 +142,11 @@ class CanvasRenderingContext2DBase : public CanvasRenderingContext, public Canva
-    void restore();
+    ExceptionOr<void> restore();
+
+    uint32_t layerCount() const { return m_layerCount; }
```
All calls to `layerCount()` actually just want to know if there are any open layers. How about we have:

```
bool hasOpenLayers() const { return m_layerCount; }
```

and replace the callers with that.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1088,9 +1111,11 @@ void CanvasRenderingContext2DBase::fillInternal(const Path& path, CanvasFillRule
-    didDraw(repaintEntireCanvas, [&] {
-        return path.fastBoundingRect();
-    });
+    if (!layerCount()) {
```
Should we put the layer count check inside `didDraw(std::optional<FloatRect>, OptionSet<DidDrawOption>)` as the first thing it does, so we don't need to repeat the check here in every drawing function?

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2490,12 +2542,16 @@ void CanvasRenderingContext2DBase::putImageData(ImageData& data, int dx, int dy,
         buffer->putPixelBuffer(data.pixelBuffer(), sourceRect, IntPoint { destOffset });
 
-    OptionSet<DidDrawOption> options; // ignore transform, shadow, clip, and post-processing
+    if (!layerCount()) {
```
We've already returned early at the start of the function if we have open layers, so we don't need this check.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2860,4 +2917,66 @@ DestinationColorSpace CanvasRenderingContext2DBase::colorSpace() const
+
+    ctx->beginTransparencyLayer(state().globalAlpha);
+
+    // Reset layer rendering states to default value: globalAlpha, globalCompositeOperation, shadowOffsetX, shadowOffsetY, shadowBlur, shadowColor.
```
Probably don't need this comment since it's reasonably clear what's happening.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2860,4 +2917,66 @@ DestinationColorSpace CanvasRenderingContext2DBase::colorSpace() const
+ExceptionOr<void> CanvasRenderingContext2DBase::endLayer()
+{
+    if (m_stateStack.isEmpty())
+        return Exception { InvalidStateError };
```
I think this should be an assertion rather than a check and throwing an exception. The state stack should never be empty.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2860,4 +2917,66 @@ DestinationColorSpace CanvasRenderingContext2DBase::colorSpace() const
+
+    auto* ctx = drawingContext();
+    if (!ctx) {
+        // FIXME: what if drawingContext() is null.
```
We generally just do nothing and return if the drawing context is empty, so I think we can remove this FIXME.

heycam commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -206,6 +207,12 @@ template<typename T> static Ref<JSON::ArrayOf<JSON::Value>> buildArrayForVector(
 
+std::optional<InspectorCanvasCallTracer::ProcessedArgument> InspectorCanvas::processArgument(BeginLayerOptions)
+{
+    // FIXME: implement this.
```
Make sure you have a bug on file for this.

heycam reviewed (approved):
r=me with comments addressed.

tuankiet65 committed:
Implement basic CanvasRenderingContext2D layers support without filters
https://bugs.webkit.org/show_bug.cgi?id=259784
rdar://113343677

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-reset-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.beginLayer-save-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.save-beginLayer-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.invalid-calls.save-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-endLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.beginLayer-save-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/layers/2d.layer.valid-calls.save-beginLayer-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::contextIs2DBaseWithUnclosedLayers const):
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::toDataURL):
(WebCore::HTMLCanvasElement::toBlob):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::convertToBlob):
* Source/WebCore/html/canvas/BeginLayerOptions.h: Added.
* Source/WebCore/html/canvas/BeginLayerOptions.idl: Added.
* Source/WebCore/html/canvas/CanvasLayers.idl: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::unwindStateStack):
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::State::State):
(WebCore::CanvasRenderingContext2DBase::realizeSavesLoop):
(WebCore::CanvasRenderingContext2DBase::restore):
(WebCore::CanvasRenderingContext2DBase::clearRect):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::createPattern):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
(WebCore::CanvasRenderingContext2DBase::beginLayer):
(WebCore::CanvasRenderingContext2DBase::endLayer):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::hasOpenLayers const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
